### PR TITLE
feat(Models): GroupLink Model 

### PR DESCRIPTION
**Pull request by MaxBittker**
this PR includes a new model + migration for group <-> resource links

the first place it will be used is GroupCommitResolution. 
in this PR, we simple start dual writing, those connections, in future PRs there will be a data migration and more code will be switched to use GroupLink

### CODE REVIEW

**Comment by getsentry-danger** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">You should update CHANGES due to the size of this PR</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">danger</a>
</p>

File: src/sentry/models/grouplink.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
+"""
+sentry.models.groupseen
```
Comment (macqueen) [inline src/sentry/models/grouplink.py]: should be grouplink

File: src/sentry/models/grouplink.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+from django.utils import timezone
+from django.utils.translation import ugettext_lazy as _
+from jsonfield import JSONField
+
+from sentry.db.models import Model, sane_repr, BoundedBigIntegerField, BoundedPositiveIntegerField
+
+
+class GroupLink(Model):
+    """
+    Link a group with an extenal resource like a commit, issue, or pull request
```
Comment (macqueen) [inline src/sentry/models/grouplink.py]: `extenal` --> `external`

File: src/sentry/models/grouplink.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    """
+    __core__ = False
+
+    class Relationship:
+        unknown = 0
+        resolves = 1
+        links = 2
```
Comment (macqueen) [inline src/sentry/models/grouplink.py]: what about calling this `references` or something? don't really care either way, just seems like a way to not overuse the word `link`

File: src/sentry/models/grouplink.py [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+        default=Relationship.links,
+        choices=((Relationship.resolves, _('Resolves')),
+                 (Relationship.links, _('Linked')), ),
+    )
+
+    data = JSONField()
+
+    datetime = models.DateTimeField(default=timezone.now, db_index=True)
+
+    unique_together = (('group_id', 'linked_type', 'linked_id'), )
```
Comment (macqueen) [inline src/sentry/models/grouplink.py]: this needs to be defined in the meta class, and you'll need to regenerate the migration to actually create the unique constraint

File: src/sentry/models/grouplink.py [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+
+    relationship = BoundedPositiveIntegerField(
+        default=Relationship.links,
+        choices=((Relationship.resolves, _('Resolves')),
+                 (Relationship.links, _('Linked')), ),
+    )
+
+    data = JSONField()
+
+    datetime = models.DateTimeField(default=timezone.now, db_index=True)
```
Comment (macqueen) [inline src/sentry/models/grouplink.py]: definitely a nit, but i think it's easier to tell what the model fields are when they're grouped together without a blank line between them

File: src/sentry/models/grouplink.py [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+        unknown = 0
+        commit = 1
+        pull = 2
+        issue = 3
+
+    group_id = BoundedBigIntegerField()
+
+    linked_type = BoundedPositiveIntegerField(
+        default=LinkedType.commit,
+        choices=((LinkedType.commit, _('Commit')),
```
Comment (macqueen) [inline src/sentry/models/grouplink.py]: i'm pretty sure `choices` are only used for django forms, which we'll probably never have for this, so not sure if this is necessary

**Review by macqueen — CHANGES_REQUESTED** [review]

File: src/sentry/models/grouplink.py [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+        resolves = 1
+        references = 2
+
+    class LinkedType:
+        unknown = 0
+        commit = 1
+        pull = 2
+        issue = 3
+
+    group_id = BoundedBigIntegerField()
```
Comment (dcramer) [inline src/sentry/models/grouplink.py]: could we also add a project_id column, as it will make some things easier in the future when we do tenants

**Review by macqueen — APPROVED** [review]
there is already a migration in master with id 0362, so you're going to need to regenerate the migration before merging. other than that, looks good!

**Pull request merged**